### ISSUE TITLE
Add further missing option to summaries (and climatic summaries

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#5310]
We currently have a Missing Tab in the Prepare > Column: Reshape > Column Summaries dialogue.

I would like a further checkbox to be added, underneath the existing materials.

Default is unchecked.  Label is **Limit on consecutive missing values**
If checked, then another label is (as the label above, but with consecutively added, i.e. 
**Summary set to missing if more than this number are missing consecutively**.  

Then an up-down (perhaps underneath the label if not enough space.  Default could be 4.

This should be easy to add and leave disabled for now.  It maybe that the R-code is easy - it is a "spell" of missing values.  Danny or David to advise.

If it can be added, this is all that is needed for R-Instat to be able to provide climate normals - and that would greatly impress Climsoft installers!

**Comment by Wycklife** [comment]
@rdstern What could be the appropriate name for the checkbox?

**Comment by rdstern** [comment]
As above:
Default is unchecked. Label is **Limit on consecutive missing values**
If checked, then another label is: 
**Summary set to missing if more than this number are missing consecutively.**

Then an up-down (perhaps underneath the label if not enough space. Default could be 4.

**Comment by shadrackkibet** [comment]
@Wycklife any update on this?

**Comment by shadrackkibet** [comment]
@dannyparsons  @volloholic I have implemented this check as a function in a similar way we did for the count of missing values. I did this initially for mean function. This is all that is needed to calculate climate normals.  Below is my implementation. Any comments?

```
na_consecutive_check <- function(x, threshold = NULL, ...){ 
  is_na_rle <- rle(is.na(x)) 
  is_na_rle$values <- is_na_rle$values & is_na_rle$lengths >= threshold 
  return(any(is_na_rle$values)) 
  }

summary_mean <- function (x, add_cols, weights="", na.rm = FALSE, trim = 0, na_type = "", threshold = NULL, ...) { 
  if( length(x)==0 || (na.rm && length(x[!is.na(x)])==0) ) return(NA) 
  else { 
    if(na.rm && na_type != "" && !na_check(x, na_type = na_type, ...) || na_consecutive_check(x, threshold = threshold, ...)) return(NA) 
    else return(mean(x, na.rm = na.rm, trim = trim)) }
}
```

**Example**
```
x <- c(NA,6,3,4,NA,NA,NA)
summary_mean(x)
[1] NA
summary_mean(x, threshold = 2, na.rm = TRUE)
[1] NA
 summary_mean(x, threshold = 4, na.rm = TRUE)
[1] 4.333333
```

**Comment by dannyparsons** [comment]
Thanks Shadrack. Could you move the code from `na_consecutive_check` into `na_check` as another option? I think we want consecutive missing to just be another type of missing check.

Then to do climate normals we need be able to require more than one type of missing check. So `na_check` needs to be modified to allow `na_type` to be a vector. If it's a vector of more than one type, then if any check fails the result should be missing. If that could be done then on the dialog the radio buttons can change to checkboxes.

**Comment by shadrackkibet** [comment]
@dannyparsons I have now implemented this as you suggested. Please have a look at it. Below is the code.

```
na_check <- function(x, na_type = c(), na_consecutive_n = NULL, na_max_n = NULL, na_max_prop = NULL, na_min_n = NULL, na_FUN = NULL, ...) {
  res <- c()
  k <- 1
  for (i in na_type) {
    if(i == "n") {
      res[k] <- summary_count_missing(x) <= na_max_n
    }
    else if(i == "prop") {
      res[k] <- (summary_count_missing(x)/summary_count(x)) <= na_max_prop
    }
    else if(i == "n_non_miss") {
      res[k] <- summary_count_non_missing(x) >= na_min_n
    }
    else if(i == "FUN") {
      res[k] <- na_FUN(x, ...)
    }
    else if(i == "c") {
      is_na_rle <- rle(is.na(x)) 
      is_na_rle$values <- is_na_rle$values & is_na_rle$lengths >= na_consecutive_n 
      res[k] <- !(any(is_na_rle$values))
    }
    else stop("Invalid na_type specified for missing values check.")
    k <-  k+1
  }
  return(length(res>1) && all(res))
}


summary_mean <- function (x, add_cols, weights = "", na.rm = FALSE, trim = 0, na_type = c(), ...) {
  if( length(x)==0 || (na.rm && length(x[!is.na(x)])==0) ) return(NA)
  else {
    if(na.rm && na_type != "" && !na_check(x, na_type = na_type, ...)) return(NA)
    else return(mean(x, na.rm = na.rm, trim = trim))
  }
}

##Example
x <- c(NA,6,3,4,NA,NA,NA)
summary_mean(x)
[1] NA
> summary_mean(x, na_type = 'c', na_consecutive_n = 3, na.rm = TRUE)
[1] NA
> summary_mean(x, na_type = 'n', na_max_n = 2, na.rm = TRUE)
[1] NA
> summary_mean(x, na_type = c('n','c'), na_max_n = 7, na_consecutive_n = 3, na.rm = TRUE)
[1] NA
> summary_mean(x, na_type = c('n','c'), na_max_n = 7, na_consecutive_n = 7, na.rm = TRUE)
[1] 4.333333
> summary_mean(x, na_type = c('n','c'), na_max_n = 3, na_consecutive_n = 3, na.rm = TRUE)
[1] NA
```

**Comment by dannyparsons** [comment]
Have discussed with Shadrack and this is almost ready to be implemented.

**Comment by rdstern** [comment]
On the missing options the third is the percentage missing.
I tried to have less than 1, but it insists on integers starting from 1.  Ideally I would like to be able to type any positive value less than 100.  But it should at least allow one decimal with 0.1 as the minimum, rather than simply integer values.

Apart from that, the missing options seem to be working, singly and also in the combinations I have tested.  Really great.
